### PR TITLE
[SPARK-18694][SS]Add StreamingQuery.explain and exception to Python and fix StreamingQueryException

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

- Add StreamingQuery.explain and exception to Python.
- Fix StreamingQueryException to not expose `OffsetSeq`. 

## How was this patch tested?

Jenkins

### CODE REVIEW

File: python/pyspark/sql/streaming.py [hunk 1]
```diff
@@ -132,6 +133,61 @@ def stop(self):
...
+        >>> bad_udf = udf(lambda x: x / 0)
+        >>> sq = sdf.select(bad_udf(col("value"))).writeStream.format('memory')\
+              .queryName('this_query').start()
+        >>> try:
+        ...     sq.processAllAvailable() # Process some data to fail the query
+        ... except:
+        ...     pass # Ignore the error as we want to test the "exception" method
+        >>> sq.exception()
+        StreamingQueryException()
+        >>> sq.stop()
```
Comment (tdas) [inline python/pyspark/sql/streaming.py]: I am not sure if such a big test should be present as doc tests.  please add this to the sql/tests.py.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala [hunk 2]
```diff
@@ -38,6 +38,13 @@ import org.apache.spark.annotation.Experimental
 class StateOperatorProgress private[sql](
     val numRowsTotal: Long,
     val numRowsUpdated: Long) {
+
+  /** The compact JSON representation of this progress. */
+  def json: String = compact(render(jsonValue))
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: is this needed?

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala [hunk 3]
```diff
@@ -103,10 +103,12 @@ class StreamingQuerySuite extends StreamTest with BeforeAndAfter with Logging {
       TestAwaitTermination(ExpectException[SparkException]),
       TestAwaitTermination(ExpectException[SparkException], timeoutMs = 2000),
       TestAwaitTermination(ExpectException[SparkException], timeoutMs = 10),
-      AssertOnQuery(
-        q => q.exception.get.startOffset.get.offsets ===
-          q.committedOffsets.toOffsetSeq(Seq(inputData), "{}").offsets,
-        "incorrect start offset on exception")
+      AssertOnQuery(q => {
+        q.exception.get.startOffset ===
+          q.committedOffsets.toOffsetSeq(Seq(inputData), "{}").toString &&
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQuerySuite.scala]: This line would conflict with my PR #16113 , you may have to fix issues after my PR gets merged.

File: sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala (hunk 2 above)
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/streaming/progress.scala]: > is this just for consistency with SourceProgress?

Yes. Other progress classes have these json methods.

**Comment by tdas** [comment]
Merging this to master and 2.1

**Comment by tdas** [comment]
Merged to master, but 2.1 failed due to mima exclude conflict. can you make another PR for 2.1 @zsxwing thanks!

**Pull request closed without merging**